src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java

```diff
@@ -399,8 +399,30 @@ public void save(DataTree dataTree,
         File snapshotFile = new File(snapDir, Util.makeSnapshotName(lastZxid));
         LOG.info("Snapshotting: 0x{} to {}", Long.toHexString(lastZxid),
                 snapshotFile);
-        snapLog.serialize(dataTree, sessionsWithTimeouts, snapshotFile, syncSnap);
-
+        try {
+            snapLog.serialize(dataTree, sessionsWithTimeouts, snapshotFile, syncSnap);
+        } catch (IOException e) {
+            if (snapshotFile.length() == 0) {
```

Comment:
this seems extra safe. i'm wondering if perhaps we should always delete the snapshot file on an exception. the snapshot file probably will not be valid if an exception occurred. right?

Reply:
>> If that sounds reasonable, I'd propose that we commit on this patch and close the hole created from this particular issue

SGTM. @breed - ok with you? If so I'll commit this.

Reply:
Created https://issues.apache.org/jira/browse/ZOOKEEPER-3115 to record this idea of deleting the snapshot file on general IOException's and we can continue the conversation there.